processes_notwindows.go

```diff
@@ -27,90 +33,127 @@ func processes() ([]ProcStat, error) {
+		procStatus := parseProcStatusFile(statusFile)
+		stat := ProcStat{ParentPID: procStatus.PPID, State: procStatus.State}
+		// get the PID from the filepath(/proc/<pid>/status) itself
+		pathParts := strings.Split(statusFilepath, string(filepath.Separator))
+		pidString := pathParts[len(pathParts)-2]
+		stat.PID, err = strconv.Atoi(pidString)
 		if err != nil {
-			log.Errorf("Failed to convert PPID(%s) to int: %s", stats[4], err.Error())
+			log.Errorf("[PROC] proc/status: failed to convert PID(%s) to int: %s", pidString, err.Error())
 		}
 
-		stat := ProcStat{PID: pid, ParentPID: ppid}
-
-		comm, err := readProcFile(getHostProc() + "/" + string(stats[0]) + "/comm")
+		commFilepath := getHostProc() + "/" + pidString + "/comm"
+		comm, err := readProcFile(commFilepath)
 		if err != nil && err != errorProcessTerminated {
-			log.Errorf("Failed to read comm(%s): %s", stats[0], err.Error())
+			log.Errorf("[PROC] failed to read comm(%s): %s", commFilepath, err.Error())
 		} else if err == nil {
 			stat.Name = string(bytes.TrimRight(comm, "\n"))
 		}
 
-		cmdline, err := readProcFile(getHostProc() + "/" + string(stats[0]) + "/cmdline")
+		cmdLineFilepath := getHostProc() + "/" + pidString + "/cmdline"
+		cmdline, err := readProcFile(cmdLineFilepath)
 		if err != nil && err != errorProcessTerminated {
-			log.Errorf("Failed to read cmdline(%s): %s", stats[0], err.Error())
+			log.Errorf("[PROC] failed to read cmdline(%s): %s", cmdLineFilepath, err.Error())
 		} else if err == nil {
 			stat.Cmdline = strings.Replace(string(bytes.TrimRight(cmdline, "\x00")), "\x00", " ", -1)
 		}
 
-		switch stats[2][0] {
-		case 'R':
-			stat.State = "running"
-		case 'S':
-			stat.State = "sleeping"
-		case 'D':
-			stat.State = "blocked"
-		case 'Z':
-			stat.State = "zombie"
-		case 'X':
-			stat.State = "dead"
-		case 'T', 't':
-			stat.State = "stopped"
-		case 'W':
-			stat.State = "paging"
-		case 'I':
-			stat.State = "idle"
-		default:
-			stat.State = "unknown"
-		}
 		procs = append(procs, stat)
 	}
 
 	return procs, nil
 }
 
+func parseProcStatusFile(b []byte) procStatus {
+	// fill default value
+	// we need non-zero values in order to check if we set them, because PPID can be 0
+	status := procStatus{
+		PPID: -1,
+	}
+
+	scanner := bufio.NewScanner(bytes.NewReader(b))
+	for scanner.Scan() {
+		fields := strings.Fields(scanner.Text())
+
+		switch strings.ToLower(fields[0]) {
+		case "ppid:":
+			var err error
+			status.PPID, err = strconv.Atoi(fields[1])
+			if err != nil {
+				log.Errorf("[PROC] proc/status: failed to convert PPID(%s) to int: %s", fields[1], err.Error())
+			}
+		case "state:":
+			// extract raw long state
+			// eg "State:	S (sleeping)"
+			if len(fields) >= 3 {
+				status.State = strings.ToLower(strings.Trim(fields[2], "()"))
+				break
+			}
+
+			// determine long state from the short one in case long one is not available
+			// eg "State:	S"
+			status.State = getProcLongState(fields[1][0])
```

Comment:
Is it possible len(fields) == 1 or len(fields[1]) == 0? Bounds check?

Reply:
The first check is enough. I use `strings.Fields()` it will not create the empty second item